### PR TITLE
Rest api extension 

### PR DESCRIPTION
**Pull request by JamesHGreen**
@RocketChat/core 

Extending rest api to allow outgoing webhooks to be created using a rest request,
Opening end points allowing interaction with Rocket.Chat via rest. 

Documentation at:
https://github.com/JamesHGreen/Rocket.Chat/wiki/Rocket.Chat-REST-API

### CODE REVIEW

File: packages/rocketchat-api/server/routes.coffee [hunk 1]
```diff
@@ -185,5 +196,195 @@ RocketChat.API.v1.addRoute 'groups.create', authRequired: true,
 			return RocketChat.API.v1.failure e.name + ': ' + e.message
 
 		return RocketChat.API.v1.success
-			group: RocketChat.models.Rooms.findOne({_id: id.rid})
+			group: RocketChat.models.Rooms.findOne({_id: id.rid})		
+
+
+#list All Private Groups		
+RocketChat.API.v1.addRoute 'privateRooms.list', authRequired: true,
```
Comment (rodrigok) [inline packages/rocketchat-api/server/routes.coffee]: We use `groups.list` and `groups.create`, why should we change the prefix here?

File: packages/rocketchat-api/server/routes.coffee [hunk 2]
```diff
@@ -185,5 +196,195 @@ RocketChat.API.v1.addRoute 'groups.create', authRequired: true,
...
+	get: ->
+		try
+			this.response.setTimeout (1000)
+			rooms = RocketChat.models.Rooms.findByType('p', { sort: { msgs:-1 } }).fetch()
```
Comment (rodrigok) [inline packages/rocketchat-api/server/routes.coffee]: We should verify some permissions, maybe we should create new permissions for these methods like `api-groups-list` that only admins have this permission

File: packages/rocketchat-api/server/routes.coffee [hunk 3]
```diff
@@ -185,5 +196,195 @@ RocketChat.API.v1.addRoute 'groups.create', authRequired: true,
...
+		return RocketChat.API.v1.success
+			list: rooms
+			
+			
+#list All Direct Groups		
+RocketChat.API.v1.addRoute 'directRoom.list', authRequired: true,
+	get: ->
+		try
+			this.response.setTimeout (1000)
+			rooms = RocketChat.models.Rooms.findByType('d', { sort: { msgs:-1 } }).fetch()
```
Comment (rodrigok) [inline packages/rocketchat-api/server/routes.coffee]: Same problem with permission here, it's open for all logged users

File: packages/rocketchat-api/server/routes.coffee [hunk 4]
```diff
@@ -185,5 +196,195 @@ RocketChat.API.v1.addRoute 'groups.create', authRequired: true,
...
+		catch e
+			return RocketChat.API.v1.failure e.name + ': ' + e.message
+		return RocketChat.API.v1.success
+			list: rooms			
+			
+			
+# addUser to a channel/private group
+RocketChat.API.v1.addRoute 'addUser', authRequired: true,
+	post: ->
+		if RocketChat.authz.hasPermission(@userId, 'bulk-create-c')
```
Comment (rodrigok) [inline packages/rocketchat-api/server/routes.coffee]: You don't need to verify this permission here, there is no relation with `bulk-create-c` and add an user to a room, the method `addUserToRoom` is doing the permission verification for you

File: packages/rocketchat-api/server/routes.coffee [hunk 5]
```diff
@@ -185,5 +196,195 @@ RocketChat.API.v1.addRoute 'groups.create', authRequired: true,
...
+
+		
+### Remove rooms
+any type i.e. channels/private/direct message room
+user must also have create-c permission
+pass room _id in body of data (  data='{"name":[" room _id"]}'  )				
+###
+RocketChat.API.v1.addRoute 'bulk/removeGroup', authRequired: true,
+	delete: ->
+		if RocketChat.authz.hasPermission(@userId, 'bulk-create-c')
```
Comment (rodrigok) [inline packages/rocketchat-api/server/routes.coffee]: We should have a permission for delete and not use the `bulk-create-c` permission

File: packages/rocketchat-api/server/routes.coffee [hunk 6]
```diff
@@ -185,5 +196,195 @@ RocketChat.API.v1.addRoute 'groups.create', authRequired: true,
...
+				
+### Retrieve integrations  
+**Note make sure you encode all channel/private rooms  because they start with '#' === '%23' 
+Example string ('http://your url here : 3000/api/v1/roomIntegrations.id/%23testRoom/list',headers={'X-User-Id':'user_token','X-Auth-Token':'user_token'})
+###
+RocketChat.API.v1.addRoute 'roomIntgrations.id/:channel/list', authRequired: true,
+	get: ->
+		try	
+			this.reaponse.setTimeout(1000)			
+			id = RocketChat.models.Integrations.find({"channel":decodeURIComponent(@urlParams.channel)}).fetch()
```
Comment (rodrigok) [inline packages/rocketchat-api/server/routes.coffee]: Permission is missing here too

**Comment by JamesHGreen** [comment]
refactor method names to match RocketChat structure.
Added correct permissions to each method,

**Comment by engelgabriel** [comment]
Thanks @JamesHGreen we will review it before the next release.

**Comment by JamesHGreen** [comment]
opened new request under #4576

**Pull request closed without merging**